packages/oauth/oauth-scopes/src/scopes/include-scope.ts

```diff
@@ -0,0 +1,147 @@
+    return permissionSet.permissions
+      .map(parseIncludedPermission, this)
+      .filter(isNonNullable)
+      .filter(isIncludedPermissionAllowed, this)
+  }
+
+  protected static readonly parser = new Parser(
+    'include',
+    {
+      nsid: {
+        multiple: false,
+        required: true,
+        validate: isNsid,
+      },
+      aud: {
+        multiple: false,
+        required: false,
+        validate: isAtprotoAudience,
+      },
+    },
+    'nsid',
+  )
+
+  static fromString(scope: string) {
+    if (!isScopeSyntaxFor(scope, 'include')) return null
+    const syntax = ScopeSyntax.fromString(scope)
+    return IncludeScope.fromSyntax(syntax)
+  }
+
+  static fromSyntax(syntax: ScopeSyntax) {
+    const result = IncludeScope.parser.parse(syntax)
+    if (!result) return null
+    return new IncludeScope(result.nsid, result.aud)
+  }
+}
+
+/**
+ * Parses a permission included thought a lexicon permission set, in the
+ * context of an `include:` scope.
+ */
+function parseIncludedPermission(
+  this: IncludeScope,
+  permission: LexPermission,
+) {
+  if (
+    permission.resource === 'rpc' &&
+    permission.inheritAud === true &&
+    permission.aud === undefined &&
+    this.aud
+  ) {
+    // "rpc" permissions can "inherit" their audience from "aud" param defined
+    // in the "include:<nsid>?aud=<audience>" scope the permission set was
+    // loaded from.
+    return parseResourcePermissionLexicon({
+      ...permission,
+      inheritAud: undefined,
+      aud: this.aud,
+    })
+  }
+
+  return parseResourcePermissionLexicon(permission)
+}
+
+/**
+ * Verifies that a permission included through a lexicon permission set is
+ * allowed in the context of the `include:` scope. This basically checks that
+ * the permission is "under" the namespace authority of the `include:` scope,
+ * and that it only contains "repo:", "rpc:", or "blob:" permissions.
+ */
+function isIncludedPermissionAllowed(
+  this: IncludeScope,
+  permission:
+    | AccountPermission
+    | BlobPermission
+    | IdentityPermission
+    | RepoPermission
+    | RpcPermission,
+): permission is RpcPermission | RepoPermission | BlobPermission {
+  if (permission instanceof RpcPermission) {
+    return permission.lxm.every(isUnderAuthority, this)
+  }
+
+  if (permission instanceof RepoPermission) {
+    return permission.collection.every(isUnderAuthority, this)
+  }
+
+  if (permission instanceof BlobPermission) {
+    return true
+  }
+
+  return false
+}
+
+/**
+ * Verifies that an nsid is under the namespace has the right authority in the
+ * context of an `include:` scope.
+ */
+function isUnderAuthority(this: IncludeScope, itemNsid: '*' | Nsid) {
+  if (itemNsid === '*') return false
+
+  const authorityNamespace = extractNsidNamespace(this.nsid)
+  const itemNamespace = extractNsidNamespace(itemNsid)
+  return (
+    itemNamespace === authorityNamespace ||
+    itemNamespace.startsWith(`${authorityNamespace}.`)
+  )
+}
```

Comment:
@bnewbold tagging just to double-check you're on the same page about this logic for determining which `lxm` and `collection` used within a permission set, based on their nsids.

Reply:
https://github.com/bluesky-social/atproto/blob/6c6cd4d186b9cea18419a85a9de18195849c3d19/packages/oauth/oauth-scopes/src/scopes/include-scope.ts#L88-L99